EVENTLISTS [readevent]

kashish2210 opened the pull request:
This PR includes the update to events.JL function so that it contains now abstract methods with a constructor, and mainly comprises the light curve function with metadata and filter as per the discussion
majorly work on this #29

lightcurve function included create_lightcurve,EventProperty, AbstractLightCurve ,rebin, calculate_errors, LightCurve, LightCurveMetadata


tagging @matteobachetti  @fjebaker  @stefanocovino

kashish2210 committed:
add abstract metod to envent reading

kashish2210 committed:
add lightcurve function

kashish2210 committed:
add some more test for events.jl as per codcov coverage

matteobachetti commented on src/events.jl:
```diff
@@ -60,26 +74,60 @@ function readevents(path; T = Float64)
                 end
+
+                # If we found both time and energy data, we can return
+                if !isempty(times) && !isempty(energies)
```
Note that most event lists do _not_ have an "energy" column, so this will keep going through the different extension needlessly. 
When you find the time column, you will also have found all the information you can find about single events (energy, PI, etc.).

matteobachetti commented on src/lightcurve.jl:
```diff
@@ -0,0 +1,309 @@
+    if method === :poisson
+        return convert.(T, sqrt.(counts))
+    elseif method === :gaussian
+        return convert.(T, sqrt.(counts .+ 1))
```
Gaussian errors need to be given by the user, estimating them like this would be wrong and, in any case, it seems to assume they are Poisson (the sqrt only makes sense in that case, and summing 1 is a quick way to give an error bar when the count rate is 0).

kashish2210 committed:
major update in event.jl

kashish2210 committed:
upadtemin error methor for light curve

fjebaker reviewed (commented):
This looks generally really good! I think there are just some redundancies and simplifications that ought to be made to help us maintain this code in the future.

We should also talk about tests and how tests should be organised. We can chat about that in the meeting tomorrow. I'd recommend not being too specific with the tests at this stage, else you'll spend most of your time maintaining the tests when you need to make a small change to how something works.

kashish2210 committed:
update

kashish2210 committed:
constructor update for sorting events validation

fjebaker reviewed (commented):
Thanks @kashish2210 this looks fine! Just some small things remaining. The code is getting quite complex, so I'd urge you to split things up into small, reusable functions in future work, but we can clean this up later in different PRs.

kashish2210 committed:
fix try and catch block and add docstring

kashish2210 commented:
@fjebaker I have made the changes Here is the summary  
## Specific catch blocks: Replaced broad catch e with specific error handling:
```
catch header_error for header reading
catch energy_error for energy column reading
catch sector_error for sector column reading
catch hdu_error for HDU access errors
catch table_error for table reading errors

```
## Reduced nesting: Simplified the logic flow by:
```
Using findfirst() instead of loops for column searching
Extracting the main HDU reading logic to reduce nesting
Using clearer conditional structures
```

kashish2210 committed:
adding updated event.jl and test cases refering #50

kashish2210 commented:
@fjebaker, I have added only events hereshallmakediffrent PR for lightcurve cause this PR getting bit complex i guess

kashish2210 committed:
add lightcurve

kashish2210 committed:
Revert "add lightcurve"

This reverts commit 34ae9e8a21553f82b461fa07757db35c2799e568.

fjebaker commented on src/events.jl:
```diff
@@ -1,85 +1,694 @@
+
+        # Get actual column names to find the correct TIME column
+        all_cols = FITSIO.colnames(selected_hdu)
+        time_col = findfirst(col -> uppercase(col) == "TIME", all_cols)
+        
+        if isnothing(time_col)
+            error("TIME column not found in HDU $hdu")
+        end
+        
+        actual_time_col = all_cols[time_col]
+        
+        # Read time column with case-insensitive option
+        time = convert(Vector{T}, read(selected_hdu, actual_time_col, case_sensitive=false))
```
```suggestion
        time = convert(Vector{T}, read(selected_hdu, "TIME", case_sensitive=false))
```

What's wrong with this? It will raise an error if no time column is found already, and is less code to reason about.

fjebaker commented on test/test_events.jl:
```diff
@@ -1,149 +1,453 @@
-            @info "Test file '$(test_filepath)' not found. Skipping this test."
-        end
-    end
+    ev = EventList(times, energies)
```
Nice (~;

fjebaker commented on test/test_events.jl:
```diff
@@ -1,149 +1,453 @@
+    @test isnothing(ev_no_energy.energies)
+    @test !has_energies(ev_no_energy)
+
+    println("✓ Basic EventList creation tests passed")
```
```suggestion
```
Not needed. The Test.jl library already provides all this information.

fjebaker commented on test/test_events.jl:
```diff
@@ -1,149 +1,453 @@
+    ev_no_energy = EventList(times_vec)
+    @test isnothing(energies(ev_no_energy))
+
+    println("✓ Accessor function tests passed")
```
```suggestion
```

fjebaker commented on test/test_events.jl:
```diff
@@ -1,149 +1,453 @@
+    @test size(ev) == (4,)
+    @test size(ev) == (length(times_vec),)
+
+    println("✓ Base interface method tests passed")
```
```suggestion
```

fjebaker commented on test/test_events.jl:
```diff
@@ -1,149 +1,453 @@
+    @test ev_extra.energies == [30.0, 40.0]
+    @test ev_extra.meta.extra_columns["INDEX"] == [3, 4]
+
+    println("✓ filter_time! function tests passed")
```
```suggestion
```

fjebaker commented on test/test_events.jl:
```diff
@@ -1,149 +1,453 @@
+    @test ev_extra.energies == [35.0, 45.0]
+    @test ev_extra.meta.extra_columns["DETX"] == [0.3, 0.4]
+
+    println("✓ filter_energy! function tests passed")
```
```suggestion
```

fjebaker commented on test/test_events.jl:
```diff
@@ -1,149 +1,453 @@
-            write(io, "This is not a FITS file")
-        end
-        @test_throws Exception readevents(invalid_file)
+    println("✓ filter_on! function tests passed")
```
```suggestion
```

fjebaker commented on test/test_events.jl:
```diff
@@ -1,149 +1,453 @@
+    # Test filter_energy error with no energies
+    @test_throws AssertionError filter_energy(e -> e > 10.0, ev_no_energy)
+
+    println("✓ Non-mutating filter function tests passed")
```
```suggestion
```

fjebaker commented on test/test_events.jl:
```diff
@@ -1,149 +1,453 @@
+    @test ev_all_pass.energies == original_energies
+    @test length(ev_all_pass) == 3
+
+    println("✓ Complex filtering scenario tests passed")
```
```suggestion
```

fjebaker commented on test/test_events.jl:
```diff
@@ -1,149 +1,453 @@
+    @test haskey(data.meta.extra_columns, "INDEX")
+    @test data.meta.extra_columns["INDEX"] == collect(1:5)
+
+    println("✓ Basic readevents tests passed")
```
```suggestion
```

fjebaker commented on test/test_events.jl:
```diff
@@ -1,149 +1,453 @@
+    @test data_hdu3.times == times
+    @test data_hdu3.energies == energies
+
+    println("✓ HDU handling tests passed")
```
```suggestion
```

fjebaker commented on test/test_events.jl:
```diff
@@ -1,149 +1,453 @@
+    @test data_pha.energies == pi_values
+    @test data_pha.meta.energy_units == "PHA"
+
+    println("✓ Alternative energy column tests passed")
```
```suggestion
```

fjebaker commented on test/test_events.jl:
```diff
@@ -1,149 +1,453 @@
+    @test isnothing(data.energies)
+    @test isnothing(data.meta.energy_units)
+
+    println("✓ Missing column tests passed")
```
```suggestion
```

fjebaker commented on test/test_events.jl:
```diff
@@ -1,149 +1,453 @@
     end
+    @test_throws Exception readevents(invalid_file)
+
+    println("✓ Error handling tests passed")
```
```suggestion
```

fjebaker commented on test/test_events.jl:
```diff
@@ -1,149 +1,453 @@
+    @test data.times == times
+    @test data.energies == energies
+
+    println("✓ Case insensitive column tests passed")
```
```suggestion
```

fjebaker reviewed (changes requested):
This is great @kashish2210 and thanks for splitting the events list and lightcurve up so that we can review them in small chunks!

Don't add anything new to this PR. If you can just address the changes I've suggested, then from my end I'm happy for this to be merged!

kashish2210 committed:
Update test/test_events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update test/test_events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update test/test_events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update test/test_events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update test/test_events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update src/events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update src/events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update src/events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update src/events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update test/test_events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update test/test_events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update test/test_events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update test/test_events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update test/test_events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update test/test_events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update test/test_events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update test/test_events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
Update test/test_events.jl

Co-authored-by: Fergus Baker <fergusbkr@gmail.com>

kashish2210 committed:
update docsting

kashish2210 commented:
Hey @matteobachetti Thank you for the feedback, I am replying your [comment](https://github.com/StingraySoftware/Stingray.jl/pull/51#discussion_r2135520170) about `function filter_energy(f, ev::EventList)` from #51 
## reason
`filter_energy!(f, ev::EventList)` - modifies the EventList in-place
`filter_energy(f, ev::EventList)` - returns a new filtered EventList (non-mutating)
This follows Julia's standard convention where functions ending with `!` are mutating (modify their arguments) and functions without `!` are non-mutating (return new objects). The same pattern applies to filter_time! vs filter_time.
However, I understand your suggestion about using a keyword argument instead. We could consolidate this into a single function like:
```julia
function filter_energy(f, ev::EventList; inplace::Bool = false)
    if inplace
        return filter_energy!(f, ev)
    else
        new_ev = deepcopy(ev)
        return filter_energy!(f, new_ev)
    end
end
```
@fjebaker @matteobachetti shall I implement this if you guys say i will add this in this PR

matteobachetti reviewed (approved):
Ok for me, I think it's a good starting point for the rest of the open PRs

fjebaker reviewed (approved):
Nice work! :+1:

kashish2210 commented:
thanks
